hoodie-client/src/main/java/com/uber/hoodie/io/HoodieMergeHandle.java

```diff
@@ -63,72 +65,97 @@ public HoodieMergeHandle(HoodieWriteConfig config,
+    while (newRecordsItr.hasNext()) {
+      HoodieRecord<T> record = newRecordsItr.next();
+      // If the first record, we need to extract some info out
+      initInternal(fileId, record);
+      keyToNewRecords.put(record.getRecordKey(), record);
+      // update the new location of the record, so we know where to find it next
+      record.setNewLocation(new HoodieRecordLocation(commitTime, fileId));
+    }
+  }
 
-    try {
-      // Load the new records in a map
-      while (newRecordsItr.hasNext()) {
-        HoodieRecord<T> record = newRecordsItr.next();
-        // If the first record, we need to extract some info out
-        if (oldFilePath == null) {
-          String latestValidFilePath = fileSystemView
-              .getLatestDataFiles(record.getPartitionPath())
-              .filter(dataFile -> dataFile.getFileId().equals(fileId))
-              .findFirst()
-              .get().getFileName();
-          writeStatus.getStat().setPrevCommit(FSUtils.getCommitTime(latestValidFilePath));
-
-          HoodiePartitionMetadata partitionMetadata = new HoodiePartitionMetadata(fs,
-              commitTime,
-              new Path(config.getBasePath()),
-              new Path(config.getBasePath(), record.getPartitionPath()));
-          partitionMetadata.trySave(TaskContext.getPartitionId());
-
-          oldFilePath = new Path(
-              config.getBasePath() + "/" + record.getPartitionPath() + "/"
-                  + latestValidFilePath);
-          String relativePath = new Path(record.getPartitionPath() + "/" + FSUtils
-              .makeDataFileName(commitTime, TaskContext.getPartitionId(), fileId)).toString();
-          newFilePath = new Path(config.getBasePath(), relativePath);
-
-          // handle cases of partial failures, for update task
-          if (fs.exists(newFilePath)) {
-            fs.delete(newFilePath, false);
-          }
-
-          logger.info(String.format("Merging new data into oldPath %s, as newPath %s",
-              oldFilePath.toString(), newFilePath.toString()));
-          // file name is same for all records, in this bunch
-          writeStatus.setFileId(fileId);
-          writeStatus.setPartitionPath(record.getPartitionPath());
-          writeStatus.getStat().setFileId(fileId);
-          writeStatus.getStat().setPath(relativePath);
+  /**
+   * Initialize the incoming records map and extract the old file path.
+   */
+  private void init(String fileId, Map<String, HoodieRecord<T>> keyToNewRecords) {
+    this.keyToNewRecords = keyToNewRecords;
+    if (keyToNewRecords.size() > 0) {
+      initInternal(fileId, keyToNewRecords.values().stream().findFirst().get());
+    }
+  }
+
+  /**
+   * Initialize write status and extract old file path
+   * @param fileId
+   * @param record
+   */
+  private void initInternal(String fileId, HoodieRecord record) {
+    if (oldFilePath == null) {
+      WriteStatus writeStatus = ReflectionUtils.loadClass(config.getWriteStatusClassName());
```

Comment:
this sort of common initialization, you can move the default constructor and make that private?

Reply:
Yeah, if we do this, then we can get rid of 2 init() methods. But it seems like the default constructor will have a lot of logic so it would then eventually make sense to anyways move it to a different method, thoughts ?